### PR TITLE
Separate rustfmt into multiple crates

### PR DESCRIPTION
**Pull request by topecongiro**
This PR separates rustfmt into multiple crates:

- rustfmt-lib
- rustfmt-emitter
- rustfmt-configuration

Since rustfmt is registered as a workspace member in the rustc repository, we cannot directly use the workspace in rustfmt. As such, we put the above crates under the rustfmt-core directory.

Unnecessarily amount of commits shows my struggle trying to publish workspace members to crates.io and to patch `ignore` crate :man_shrugging: Sorry about that!

### CODE REVIEW

**Comment by calebcartwright** [comment]
Awesome! One big step closer to rustfmt 2.x!

Regarding the [patched version](https://crates.io/crates/rustfmt_ignore) of `ignore`, is that related to the crossbeam issue and do you anticipate using the patched version long term or just until `ignore` itself is updated (ref https://github.com/BurntSushi/ripgrep/issues/1423)?

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -62,3 +62,9 @@ jobs:
     - name: 'test ignored'
       run: cargo test -- --ignored
       shell: cmd
+    - name: 'test rustfmt-core'
+      run: cargo test --manifest-path rustfmt-core/Cargo.toml
+      shell: cmd
+    - name: 'test rustfmt-core ignored'
+      run: cargo test --manifest-path rustfmt-core/Cargo.toml -- --ignored
+      shell: cmd
```
Comment (calebcartwright) [inline .github/workflows/windows.yml]: Another enhacement we could consider making with GHA in the future would be to set up some additional workflows that have directory/crate specific triggers. That way if the only changes are to the emitter crate, our CI will only have to build the emitter crate

File: rustfmt-core/rustfmt-lib/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+[package]
+name = "rustfmt_lib"
+version = "1.0.0"
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-lib/Cargo.toml]: What'd you have in mind for versioning strategy between these different crates and the main rustfmt bin folks will get via rustup? Will they each be versioned independently, should the lib and bin versions be kepty in sync, etc.?

File: rustfmt-core/rustfmt-lib/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+authors = ["topecongiro <seuchida@gmail.com>"]
+edition = "2018"
+description = "Rustfmt as a library"
+license = "Apache-2.0/MIT"
+repository = "https://github.com/rust-lang/rustfmt"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (calebcartwright) [inline rustfmt-core/rustfmt-lib/Cargo.toml]: Can probably remove the cargo-generated lines like this right?

File: rustfmt-core/rustfmt-lib/Cargo.toml (hunk 2 above)
Comment (topecongiro) [reply rustfmt-core/rustfmt-lib/Cargo.toml]: The crates in rustfmt-core will be versioned differently from binaries (e.g., rustfmt 3.0 may depend on rustfmt-core 4.0, rustfmt-emitter 2.0, and rustfmt-configuration 3.0).

File: rustfmt-core/rustfmt-lib/Cargo.toml (hunk 3 above)
Comment (topecongiro) [reply rustfmt-core/rustfmt-lib/Cargo.toml]: Yeah, you are right. Good catch, thank you!

**Comment by topecongiro** [comment]
> Regarding the patched version of ignore, is that related to the crossbeam issue and do you anticipate using the patched version long term or just until ignore itself is updated (ref BurntSushi/ripgrep#1423)?

Yes, using rustfmt-ignore is a temporal workaround.

**Comment by BurntSushi** [comment]
FYI, `ignore 0.4.11` on crates.io now uses `crossbeam-channel 0.4`.

**Comment by calebcartwright** [comment]
Awesome, thanks @BurntSushi!

**Pull request merged**